File: native-ios/Solvio/Core/ScanQueueManager.swift
import Foundation
import SwiftUI
import UIKit

/// Background queue for receipt OCR uploads.
///
/// **Why this exists:** the previous flow was strictly one-at-a-time — pick
/// a single image, watch a full-screen spinner, confirm in a sheet, repeat.
/// Users with a stack of 10 receipts had to babysit the app for minutes.
/// Now: pick all 10 at once, get a floating progress chip, keep using the
/// app while uploads + OCR run in parallel.
///
/// **Behaviour:**
///   - User adds N images → N `ScanQueueItem`s pile up with status `.pending`.
///   - We process up to `maxConcurrent` items in parallel, transitioning
///     each through `.uploading → .processing → .saved` (or `.failed`).
///   - Server-side, the OCR endpoint persists the receipt as soon as it's
///     parsed — so by the time `.saved` fires, the receipt is already in
///     the user's archive. No follow-up confirm step needed; the user can
///     tap the floating chip to jump to receipt details if they want to
///     edit anything.
///   - On every `.saved`, we ping `AppDataStore` to refresh receipts +
///     dashboard slices so newly-scanned items appear immediately.
@MainActor
final class ScanQueueManager: ObservableObject {

    // MARK: Item

    enum ScanStatus: Equatable {
        case pending
        case uploading
        case processing
        case saved
        case failed(String)

        var isTerminal: Bool {
            switch self {
            case .saved, .failed: return true
            default: return false
            }
        }
    }

    struct ScanQueueItem: Identifiable, Equatable {
        let id: UUID
        let thumbnail: UIImage
        let imageData: Data
        let filename: String
        let createdAt: Date
        var status: ScanStatus
        var receiptId: String?
        var vendor: String?
        var total: Double?
        var currency: String?

        static func == (lhs: ScanQueueItem, rhs: ScanQueueItem) -> Bool {
            lhs.id == rhs.id && lhs.status == rhs.status && lhs.receiptId == rhs.receiptId
        }
    }

    // MARK: State

    @Published private(set) var items: [ScanQueueItem] = []

    /// Max parallel uploads. Tuned for typical mobile bandwidth — too many
    /// concurrent OCR calls bottleneck on the server side anyway.
    private let maxConcurrent = 2
    private let store: AppDataStore
    /// Set once from `SolvioApp` so failure messages get localized rather
    /// than leaking raw `error.localizedDescription` to the user.
    weak var locale: AppLocale?

    // `nonisolated` so the detached compression task in `enqueue` can read
    // them without main-actor hopping. They're immutable constants —
    // perfectly safe to read from any thread.
    //
    // Vercel serverless caps request bodies at ~4.5 MB. Anything larger
    // gets rejected with HTTP 413 before our route even runs, so we keep
    // the first-pass upload comfortably under that.
    nonisolated private static let maxPixelDimension: CGFloat = 1600
    nonisolated private static let maxUploadBytes = 4 * 1024 * 1024 // 4 MB — Vercel limit ~4.5 MB
    /// Used when the first upload was already too large for Vercel — we
    /// downscale aggressively and try one more time before marking failed.
    nonisolated private static let retryPixelDimension: CGFloat = 1024
    nonisolated private static let retryUploadBytes = 2 * 1024 * 1024 // 2 MB

    init(store: AppDataStore) {
        self.store = store
    }

    // MARK: - Public API

    /// Adds images to the queue and immediately starts processing.
    /// All compression / resizing is done off-main so the picker dismisses
    /// without jank.
    func enqueue(_ images: [UIImage]) {
        guard !images.isEmpty else { return }
        Task.detached(priority: .userInitiated) { [weak self] in
            var prepared: [ScanQueueItem] = []
            for image in images {
                let resized = Self.resize(image, maxDimension: Self.maxPixelDimension)
                guard let data = Self.compressProgressive(resized, maxBytes: Self.maxUploadBytes) else { continue }
                let thumb = Self.thumbnail(from: resized, maxSide: 96)
                prepared.append(ScanQueueItem(
                    id: UUID(),
                    thumbnail: thumb,
                    imageData: data,
                    filename: "receipt-\(Int(Date().timeIntervalSince1970))-\(UUID().uuidString.prefix(6)).jpg",
                    createdAt: Date(),
                    status: .pending
                ))
            }
            await self?.appendAndPump(prepared)
        }
    }

    /// Retry a failed item.
    func retry(id: UUID) {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return }
        guard case .failed = items[idx].status else { return }
        items[idx].status = .pending
        pump()
    }

    /// Drop a single item from the queue (only allowed for terminal states).
    func remove(id: UUID) {
        items.removeAll { $0.id == id && $0.status.isTerminal }
    }

    /// Clear every saved + failed item, leaving in-flight work alone.
    func clearCompleted() {
        items.removeAll { $0.status.isTerminal }
    }

    // MARK: - Derived state for the UI

    var inFlightCount: Int {
        items.filter { !$0.status.isTerminal }.count
    }

    var savedCount: Int {
        items.filter { $0.status == .saved }.count
    }

    var failedCount: Int {
        items.filter { if case .failed = $0.status { return true } else { return false } }.count
    }

    var hasActivity: Bool {
        !items.isEmpty
    }

    var hasInFlight: Bool {
        inFlightCount > 0
    }

    /// Aggregate progress 0…1 — count of done / total.
    var progress: Double {
        guard !items.isEmpty else { return 0 }
        let done = items.filter(\.status.isTerminal).count
        return Double(done) / Double(items.count)
    }

    // MARK: - Private

    private func appendAndPump(_ newItems: [ScanQueueItem]) {
        items.append(contentsOf: newItems)
        pump()
    }

    /// Spin up tasks until we hit the concurrency cap.
    private func pump() {
        let inFlight = items.filter { $0.status == .uploading || $0.status == .processing }.count
        var slots = max(0, maxConcurrent - inFlight)
        guard slots > 0 else { return }
        for idx in items.indices where slots > 0 {
            if items[idx].status == .pending {
                slots -= 1
                let id = items[idx].id
                items[idx].status = .uploading
                Task { [weak self] in
                    await self?.processItem(id: id)
                }
            }
        }
    }

    private func processItem(id: UUID) async {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return }
        let item = items[idx]
        do {
            // Once we hand the data to the network, we're effectively
            // "uploading" until the server replies.
            items[idx].status = .uploading
            Self.logScanAttempt(stage: "primary", bytes: item.imageData.count, filename: item.filename)
            let response = try await ReceiptsRepo.scan(
                imageData: item.imageData,
                filename: item.filename
            )
            // Consider OCR work done; either succeeded or failed parsing.
            if let success = response.firstSuccess, let receiptId = success.receiptId {
                if let nidx = items.firstIndex(where: { $0.id == id }) {
                    items[nidx].status = .saved
                    items[nidx].receiptId = receiptId
                    items[nidx].vendor = success.data?.merchant
                    items[nidx].total = success.data?.total
                    items[nidx].currency = success.data?.currency
                }
                // New receipt → invalidate caches so dashboards refresh.
                store.didMutateReceipts()
            } else {
                let msg = response.results.first?.error
                    ?? response.results.first?.message
                    ?? localized("receipts.noReceiptDetected")
                if let nidx = items.firstIndex(where: { $0.id == id }) {
                    items[nidx].status = .failed(msg)
                }
            }
        } catch ApiError.cancelled {
            // Underlying URLSession task got cancelled — requeue as
            // pending so the next pump retries instead of marking the
            // image as a permanent failure. Happens rarely in practice
            // (the queue lives at app root, not bound to view lifecycle).
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .pending
            }
        } catch ApiError.payloadTooLarge {
            // Vercel rejected the body (~4.5 MB cap). Try once more with
            // aggressive resize/compress, then mark failed if still too big.
            await retryWithAggressiveCompression(id: id, original: item)
        } catch let apiError as ApiError {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(friendlyMessage(for: apiError))
            }
        } catch {
            #if DEBUG
            print("[ScanQueue] Unexpected error: \(error)")
            #endif
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(localized("errors.unknown"))
            }
        }
        pump()
    }

    private func retryWithAggressiveCompression(id: UUID, original: ScanQueueItem) async {
        // Decode the already-compressed JPEG, resize down hard, recompress.
        guard let image = UIImage(data: original.imageData) else {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(localized("errors.payloadTooLarge"))
            }
            return
        }
        let resized = Self.resize(image, maxDimension: Self.retryPixelDimension)
        guard let jpeg = Self.compressProgressive(resized, maxBytes: Self.retryUploadBytes) else {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(localized("errors.payloadTooLarge"))
            }
            return
        }
        Self.logScanAttempt(stage: "retry", bytes: jpeg.count, filename: original.filename)
        do {
            let response = try await ReceiptsRepo.scan(
                imageData: jpeg,
                filename: original.filename
            )
            if let success = response.firstSuccess, let receiptId = success.receiptId {
                if let nidx = items.firstIndex(where: { $0.id == id }) {
                    items[nidx].status = .saved
                    items[nidx].receiptId = receiptId
                    items[nidx].vendor = success.data?.merchant
                    items[nidx].total = success.data?.total
                    items[nidx].currency = success.data?.currency
                }
                store.didMutateReceipts()
            } else {
                let msg = response.results.first?.error
                    ?? response.results.first?.message
                    ?? localized("receipts.noReceiptDetected")
                if let nidx = items.firstIndex(where: { $0.id == id }) {
                    items[nidx].status = .failed(msg)
                }
            }
        } catch ApiError.cancelled {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .pending
            }
        } catch ApiError.payloadTooLarge {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(localized("errors.payloadTooLarge"))
            }
        } catch let apiError as ApiError {
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(friendlyMessage(for: apiError))
            }
        } catch {
            #if DEBUG
            print("[ScanQueue] Retry unexpected error: \(error)")
            #endif
            if let nidx = items.firstIndex(where: { $0.id == id }) {
                items[nidx].status = .failed(localized("errors.unknown"))
            }
        }
    }

    /// Map an `ApiError` to the same user-friendly localized strings used
    /// by `ScanFlowViewModel`. Falls back to English if `locale` wasn't
    /// wired up — never leaks `localizedDescription` to the user.
    private func friendlyMessage(for error: ApiError) -> String {
        switch error {
        case .invalidURL: return localized("errors.unknown")
        case .transport: return localized("errors.network")
        case .decoding: return localized("errors.serverUnexpected")
        case .unauthorized: return localized("errors.sessionExpired")
        case .forbidden: return localized("errors.forbidden")
        case .notFound: return localized("errors.notFound")
        case .rateLimited: return localized("errors.rateLimited")
        case .timeout: return localized("errors.timeout")
        case .noConnection: return localized("errors.network")
        case .server(let status, _) where status >= 500: return localized("errors.serverDown")
        case .server: return localized("errors.serverUnexpected")
        case .payloadTooLarge: return localized("errors.payloadTooLarge")
        case .cancelled: return localized("errors.cancelled")
        case .unknown: return localized("errors.unknown")
        }
    }

    private func localized(_ key: String) -> String {
        locale?.t(key) ?? L10n.strings[.en]?[key] ?? key
    }

    private static func logScanAttempt(stage: String, bytes: Int, filename: String) {
        #if DEBUG
        let kb = Double(bytes) / 1024.0
        print(String(format: "[ScanQueue] %@: %.1f KB (%@)", stage, kb, filename))
        #endif
    }

    // MARK: - Image utilities

    nonisolated static func resize(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size
        guard max(size.width, size.height) > maxDimension else { return image }
        let scale = maxDimension / max(size.width, size.height)
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }

    nonisolated static func compressProgressive(_ image: UIImage, maxBytes: Int) -> Data? {
        for quality: CGFloat in [0.75, 0.55, 0.35, 0.20] {
            if let data = image.jpegData(compressionQuality: quality), data.count <= maxBytes {
                return data
            }
        }
        return image.jpegData(compressionQuality: 0.15)
    }

    /// Tiny thumbnail used by the floating progress chip.
    nonisolated static func thumbnail(from image: UIImage, maxSide: CGFloat) -> UIImage {
        let size = image.size
        let scale = maxSide / max(size.width, size.height)
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }
}

File: native-android/app/src/main/java/com/programo/solvio/core/scanqueue/ScanQueueManager.kt
package com.programo.solvio.core.scanqueue

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.programo.solvio.core.locale.AppLocale
import com.programo.solvio.core.network.ApiError
import com.programo.solvio.core.network.ReceiptsRepo
import com.programo.solvio.core.store.AppDataStore
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream
import java.util.UUID

/** Mirror of iOS [ScanQueueManager]. Background queue for receipt OCR uploads. */
class ScanQueueManager(
    private val store: AppDataStore,
) {
    sealed interface ScanStatus {
        data object Pending : ScanStatus
        data object Uploading : ScanStatus
        data object Processing : ScanStatus
        data object Saved : ScanStatus
        data class Failed(val message: String) : ScanStatus

        val isTerminal: Boolean
            get() = this is Saved || this is Failed
    }

    data class ScanQueueItem(
        val id: String = UUID.randomUUID().toString(),
        val thumbnail: Bitmap,
        val imageBytes: ByteArray,
        val filename: String,
        val createdAtMs: Long = System.currentTimeMillis(),
        val status: ScanStatus = ScanStatus.Pending,
        val receiptId: String? = null,
        val vendor: String? = null,
        val total: Double? = null,
        val currency: String? = null,
    ) {
        override fun equals(other: Any?): Boolean =
            other is ScanQueueItem && other.id == id && other.status == status && other.receiptId == receiptId
        override fun hashCode(): Int = id.hashCode()
    }

    private val scope: CoroutineScope = MainScope()
    private val maxConcurrent = 2
    var locale: AppLocale? = null

    var items by mutableStateOf<List<ScanQueueItem>>(emptyList()); private set

    val inFlightCount: Int get() = items.count { !it.status.isTerminal }
    val savedCount: Int get() = items.count { it.status == ScanStatus.Saved }
    val failedCount: Int get() = items.count { it.status is ScanStatus.Failed }
    val hasActivity: Boolean get() = items.isNotEmpty()
    val hasInFlight: Boolean get() = inFlightCount > 0
    val progress: Double get() {
        if (items.isEmpty()) return 0.0
        return items.count { it.status.isTerminal } / items.size.toDouble()
    }

    fun enqueue(images: List<Bitmap>) {
        if (images.isEmpty()) return
        scope.launch(Dispatchers.Default) {
            val prepared = images.mapNotNull { src ->
                val resized = resize(src, MAX_PIXEL_DIMENSION)
                val data = compressProgressive(resized, MAX_UPLOAD_BYTES) ?: return@mapNotNull null
                val thumb = thumbnail(resized, 96)
                ScanQueueItem(
                    thumbnail = thumb,
                    imageBytes = data,
                    filename = "receipt-${System.currentTimeMillis()}-${UUID.randomUUID().toString().take(6)}.jpg",
                )
            }
            scope.launch {
                items = items + prepared
                pump()
            }
        }
    }

    fun retry(id: String) {
        items = items.map {
            if (it.id == id && it.status is ScanStatus.Failed) it.copy(status = ScanStatus.Pending) else it
        }
        pump()
    }

    fun remove(id: String) {
        items = items.filterNot { it.id == id && it.status.isTerminal }
    }

    fun clearCompleted() {
        items = items.filterNot { it.status.isTerminal }
    }

    private fun pump() {
        val inFlight = items.count { it.status == ScanStatus.Uploading || it.status == ScanStatus.Processing }
        var slots = (maxConcurrent - inFlight).coerceAtLeast(0)
        if (slots == 0) return
        items.forEach { item ->
            if (slots > 0 && item.status == ScanStatus.Pending) {
                slots -= 1
                updateItem(item.id) { it.copy(status = ScanStatus.Uploading) }
                scope.launch { processItem(item.id) }
            }
        }
    }

    private suspend fun processItem(id: String) {
        val item = items.firstOrNull { it.id == id } ?: return
        try {
            updateItem(id) { it.copy(status = ScanStatus.Uploading) }
            val response = ReceiptsRepo.scan(item.imageBytes, filename = item.filename)
            val success = response.firstSuccess
            val rid = success?.receiptId
            if (success != null && rid != null) {
                updateItem(id) {
                    it.copy(
                        status = ScanStatus.Saved,
                        receiptId = rid,
                        vendor = success.data?.merchant,
                        total = success.data?.total,
                        currency = success.data?.currency,
                    )
                }
                store.didMutateReceipts()
            } else {
                val msg = response.results.firstOrNull()?.error
                    ?: response.results.firstOrNull()?.message
                    ?: localized("receipts.noReceiptDetected")
                updateItem(id) { it.copy(status = ScanStatus.Failed(msg)) }
            }
        } catch (_: ApiError.Cancelled) {
            updateItem(id) { it.copy(status = ScanStatus.Pending) }
        } catch (_: ApiError.PayloadTooLarge) {
            retryAggressive(id, item)
        } catch (_: CancellationException) {
            updateItem(id) { it.copy(status = ScanStatus.Pending) }
        } catch (e: ApiError) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(friendly(e))) }
        } catch (e: Throwable) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(localized("errors.unknown"))) }
        }
        pump()
    }

    private suspend fun retryAggressive(id: String, original: ScanQueueItem) {
        val bmp = BitmapFactory.decodeByteArray(original.imageBytes, 0, original.imageBytes.size)
        if (bmp == null) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(localized("errors.payloadTooLarge"))) }
            return
        }
        val resized = resize(bmp, RETRY_PIXEL_DIMENSION)
        val jpeg = compressProgressive(resized, RETRY_UPLOAD_BYTES)
        if (jpeg == null) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(localized("errors.payloadTooLarge"))) }
            return
        }
        try {
            val response = ReceiptsRepo.scan(jpeg, filename = original.filename)
            val success = response.firstSuccess
            val rid = success?.receiptId
            if (success != null && rid != null) {
                updateItem(id) {
                    it.copy(
                        status = ScanStatus.Saved,
                        receiptId = rid,
                        vendor = success.data?.merchant,
                        total = success.data?.total,
                        currency = success.data?.currency,
                    )
                }
                store.didMutateReceipts()
            } else {
                val msg = response.results.firstOrNull()?.error
                    ?: response.results.firstOrNull()?.message
                    ?: localized("receipts.noReceiptDetected")
                updateItem(id) { it.copy(status = ScanStatus.Failed(msg)) }
            }
        } catch (_: ApiError.Cancelled) {
            updateItem(id) { it.copy(status = ScanStatus.Pending) }
        } catch (_: ApiError.PayloadTooLarge) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(localized("errors.payloadTooLarge"))) }
        } catch (e: ApiError) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(friendly(e))) }
        } catch (_: CancellationException) {
            updateItem(id) { it.copy(status = ScanStatus.Pending) }
        } catch (_: Throwable) {
            updateItem(id) { it.copy(status = ScanStatus.Failed(localized("errors.unknown"))) }
        }
    }

    private fun updateItem(id: String, transform: (ScanQueueItem) -> ScanQueueItem) {
        items = items.map { if (it.id == id) transform(it) else it }
    }

    private fun friendly(e: ApiError): String = when (e) {
        ApiError.InvalidUrl -> localized("errors.unknown")
        is ApiError.Transport -> localized("errors.network")
        is ApiError.Decoding -> localized("errors.serverUnexpected")
        ApiError.Unauthorized -> localized("errors.sessionExpired")
        ApiError.Forbidden -> localized("errors.forbidden")
        ApiError.NotFound -> localized("errors.notFound")
        ApiError.RateLimited -> localized("errors.rateLimited")
        ApiError.Timeout -> localized("errors.timeout")
        ApiError.NoConnection -> localized("errors.network")
        is ApiError.Server -> if (e.status >= 500) localized("errors.serverDown") else localized("errors.serverUnexpected")
        ApiError.PayloadTooLarge -> localized("errors.payloadTooLarge")
        ApiError.Cancelled -> localized("errors.cancelled")
        ApiError.Unknown -> localized("errors.unknown")
    }

    private fun localized(key: String): String = locale?.t(key) ?: key

    companion object {
        private const val MAX_PIXEL_DIMENSION = 1600
        private const val MAX_UPLOAD_BYTES = 4 * 1024 * 1024
        private const val RETRY_PIXEL_DIMENSION = 1024
        private const val RETRY_UPLOAD_BYTES = 2 * 1024 * 1024

        fun resize(image: Bitmap, maxDimension: Int): Bitmap {
            val w = image.width
            val h = image.height
            val largest = maxOf(w, h)
            if (largest <= maxDimension) return image
            val scale = maxDimension.toFloat() / largest
            val newW = (w * scale).toInt().coerceAtLeast(1)
            val newH = (h * scale).toInt().coerceAtLeast(1)
            return Bitmap.createScaledBitmap(image, newW, newH, true)
        }

        fun compressProgressive(image: Bitmap, maxBytes: Int): ByteArray? {
            for (quality in intArrayOf(75, 55, 35, 20)) {
                val out = ByteArrayOutputStream()
                image.compress(Bitmap.CompressFormat.JPEG, quality, out)
                val bytes = out.toByteArray()
                if (bytes.size <= maxBytes) return bytes
            }
            val out = ByteArrayOutputStream()
            image.compress(Bitmap.CompressFormat.JPEG, 15, out)
            return out.toByteArray()
        }

        fun thumbnail(image: Bitmap, maxSide: Int): Bitmap = resize(image, maxSide)
    }
}

val LocalScanQueue = compositionLocalOf<ScanQueueManager> { error("ScanQueueManager not provided") }
